gprofiler/perf.py

```diff
@@ -55,29 +55,23 @@ def _run_perf(self, filename_base: str, dwarf=False):
                 [resource_path("perf")] + buildid_args + ["record"] + args + ["--", "sleep", str(self._duration)],
                 stop_event=self._stop_event,
             )
-            with open(parsed_path, "w") as f:
-                run_process(
-                    [resource_path("perf")] + buildid_args + ["script", "-F", "+pid", "-i", record_file.name], stdout=f
-                )
-            return parsed_path
+            perf_script_result = run_process(
+                [resource_path("perf")] + buildid_args + ["script", "-F", "+pid", "-i", record_file.name]
+            )
+            return perf_script_result.stdout.decode('utf8')
 
-    def snapshot(self) -> Iterable[Mapping[str, str]]:
+    def snapshot(self) -> Dict[int, List[Dict[str, str]]]:
         free_disk = psutil.disk_usage(self._storage_dir).free
         if free_disk < 4 * 1024 * 1024:
             raise Exception(f"Free disk space: {free_disk}kb. Skipping perf!")
 
         logger.info("Running global perf...")
-        record_path = self._run_perf("global")
-        logger.info("Finished running global perf")
-        return parse_perf_script(Path(record_path).read_text())
+        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+            fp_future = executor.submit(self._run_perf, False) if self._fp_perf else None
+            dwarf_future = executor.submit(self._run_perf, True) if self._dwarf_perf else None
+        fp_perf = fp_future.result() if fp_future is not None else None
+        dwarf_perf = dwarf_future.result() if dwarf_future is not None else None
 
-        # TODO: run dwarf in parallel, after supporting it in merge.py
-        # Alternatively: fix golang dwarf problems and the run just dwarf
-
-        # if get_cpu_count() > 32:
```

Comment:
Place some limit on the dwarf frequency - it's much heavier than FP, if someone runs gProfiler with e.g freq 1k~ and they have lots of cores, the effect will be substantial.